### PR TITLE
kiagnose, checkup: Use k8s client-go `Until` tool for watching the job

### PR DESCRIPTION
**Pull request by EdDev**
Kiagnose creates a checkup job and waits for it to finish.
The current implementation uses a watcher with a self curated logic to
process the events received from it.

The client-go tooling module has a `watch` package [1] that exposes the
function `Until`, implementing all needed mechanics that are required to
process watchers events.

By using this tool, it is now possible to drop our own limited
implementation.
It is worth noting that `Until` also handles possible connectivity
issues between the client and the server, increasing stability in case
of connectivity flakiness.

The change is following the work done for the namespace deletion [2] but
includes an improvements for early termination:
- In case a DELETED event type is detected, fail the waiting earlier.

[1] https://pkg.go.dev/k8s.io/client-go/tools/watch
[2] https://github.com/kiagnose/kiagnose/pull/84

Note: The early exit from the waiting does not include the ADDED event, as I am unsure if it will not be received as well (given we provide the resource version for the initial creation).

### CODE REVIEW

**Comment by RamLavi** [comment]
@EdDev I can't fully understand the commit message. can you rephrase? Notice:
- typo: "It it".
- You are talking about the go-tooling but not linking it.
- You are referencing to another case "It it similar to the one..." - it is not understood unless one reviewed it that case, you need to either link to that "similar case" or make the commit justification understood standalone.

**Review by RamLavi — CHANGES_REQUESTED** [review]
nit

File: kiagnose/internal/checkup/job/job.go [hunk 1]
```diff
@@ -43,42 +45,49 @@ func Create(client kubernetes.Interface, job *batchv1.Job) (*batchv1.Job, error)
...
 
-	finishedJob, err := waitForJob(ctx, jobWatcher)
+	event, err := k8swatchtools.Until(ctx, job.ResourceVersion, w, func(event k8swatch.Event) (bool, error) {
+		j, ok := event.Object.(*batchv1.Job)
+		if !ok {
+			return false, nil
+		}
+
+		switch event.Type {
+		case k8swatch.Deleted:
```
Comment (ormergi) [inline kiagnose/internal/checkup/job/job.go:71]: I think its odd that on Deleted event it will exit with `unexpected error` but for other types of event it return silently,
other events are not expected as well isnt it?
Also I think `ctx.Done` should be taken in account.

How about changing this switch to include:
case for `Added` & `Modified` events, case for `ctx.Done` and  return `unexpected event` error at `default`?

**Review by qinqon — DISMISSED** [review]

File: kiagnose/internal/checkup/job/job.go [hunk 2]
```diff
@@ -43,42 +45,49 @@ func Create(client kubernetes.Interface, job *batchv1.Job) (*batchv1.Job, error)
...
+
+	const JobNameLabel = "job-name"
+	jobLabel := k8slabels.Set{JobNameLabel: job.Name}
+	w := &cache.ListWatch{
+		WatchFunc: func(options metav1.ListOptions) (k8swatch.Interface, error) {
+			if options.LabelSelector != "" {
+				options.LabelSelector = "," + jobLabel.String()
+			} else {
+				options.LabelSelector = jobLabel.String()
+			}
```
[the remark this answers is not in the document]
Comment (EdDev) [reply kiagnose/internal/checkup/job/job.go:59]: Although it is not a must, I find this safer for potential future changes.
it is true that the `Until` function is not passing in any LabelSelector, but this is an implementation detail you rely on based on reading its code. It does pass in other options.

I think we should not rely on these implementation details and assume nothing about if the label selectors are already there or not.

File: kiagnose/internal/checkup/job/job.go (hunk 1 above)
Comment (EdDev) [reply kiagnose/internal/checkup/job/job.go:71]: I only optimized for the parts I am sure about, I am not 100% sure about `ADDED`, because the RV we watch from is the one received from the creation operation. Does in include this initial `ADDED` event? I do not know.
I guess a following PR can try to add it, explaining why it is correct.

Regarding the `ctx.Done`, that is not required as the tooling is taking care of it. The whole point of using the `watch` tooling is to offload these mechanics to it.

**Review by qinqon — CHANGES_REQUESTED** [review]
a proposal, is not a must

**Comment by EdDev** [comment]
change: Passed to `Until` a function instead of defining an anonymous one inline (per review request).

File: kiagnose/internal/checkup/checkup_test.go [hunk 3]
```diff
@@ -362,35 +364,35 @@ func TestCheckupRunShouldFailWhen(t *testing.T) {
...
 		testClient := newNormalizedFakeClientset()
-		testClient.injectJobWatchError(expectedErr)
 		testClient.injectResourceVersionUpdateOnNamespaceCreation()
 		testClient.injectWatchWithNamespaceDeleteEvent()
 
 		testCheckup := checkup.New(testClient, &config.Config{Image: testImage, Timeout: testTimeout}, nameGeneratorStub{})
 
 		assert.NoError(t, testCheckup.Setup())
-		assert.ErrorContains(t, testCheckup.Run(), expectedErr)
+		assert.ErrorContains(t, testCheckup.Run(), "initial RV \"\" is not supported")
```
Comment (RamLavi) [inline kiagnose/internal/checkup/checkup_test.go:374]: is there no type of error like this? if possible it is better to use their errors, because it is not guaranteed that the wording will stay the same in future bumps, thus braking this test.

**Review by RamLavi — CHANGES_REQUESTED** [review]
I like this new tool! Added some comments.

**Comment by EdDev** [comment]
@RamLavi , I interpreted silence as ok to take it as is. If you think that point about the test merits a helper, please suggest it in a PR and I will take it.

File: kiagnose/internal/checkup/checkup_test.go [hunk 4]
```diff
@@ -312,6 +311,7 @@ func TestCheckupRunShouldCreateAJob(t *testing.T) {
 			actualJob, err := testClient.BatchV1().Jobs(checkupNamespaceName).Get(context.Background(), checkup.JobName, metav1.GetOptions{})
 			assert.NoError(t, err)
 
+			actualJob.ResourceVersion = ""
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply kiagnose/internal/checkup/checkup_test.go:314]: I will insist! but I'll put it i another PR :)

File: kiagnose/internal/checkup/checkup_test.go (hunk 3 above)
Comment (RamLavi) [reply kiagnose/internal/checkup/checkup_test.go:374]: if you don't have  a predefined error, then I would let it go. odds are slim they they will change it...
